ui/src/components/Publication/RelatedPublications/VotingArea/index.tsx

```diff
@@ -0,0 +1,126 @@
+import React, { useState } from 'react';
+import * as OutlineIcons from '@heroicons/react/24/outline';
+
+import * as api from '@/api';
+import * as Components from '@/components';
+import * as Config from '@/config';
+import * as I from '@/interfaces';
+import * as Stores from '@/stores';
+import * as Types from '@/types';
+
+type Props = {
+    suggestedFromPublication: {
+        id: string;
+        title: string;
+    };
+    crosslink: I.Crosslink;
+    vote: boolean | null;
+};
+
+const VotingArea: React.FC<Props> = (props): React.ReactElement => {
+    const user = Stores.useAuthStore((state: Types.AuthStoreType) => state.user);
+    const [userHasUpvoted, setUserHasUpvoted] = useState<boolean>(props.vote === true);
+    const [userHasDownVoted, setUserHasDownVoted] = useState<boolean>(props.vote === false);
+    const [upvotes, setUpvotes] = useState<number>(props.crosslink.upvotes);
+    const [downvotes, setDownvotes] = useState<number>(props.crosslink.downvotes);
+    const paragraphClasses = 'mt-4 text-sm text-white-100 dark:text-grey-50';
+
+    const handleVote = async (vote: boolean) => {
+        const resettingUpvote = vote && userHasUpvoted;
+        const resettingDownVote = !vote && userHasDownVoted;
+        const resettingVote = resettingUpvote || resettingDownVote;
+        const neutralToVote = !userHasUpvoted && !userHasDownVoted;
+        const downToUp = vote && userHasDownVoted;
+        const upToDown = !vote && userHasUpvoted;
+        const changingVote = neutralToVote || downToUp || upToDown;
+
+        if (resettingVote) {
+            if (resettingUpvote) {
+                setUserHasUpvoted(false);
+                setUpvotes((prev) => prev - 1);
+            } else if (resettingDownVote) {
+                setUserHasDownVoted(false);
+                setDownvotes((prev) => prev - 1);
+            }
+            try {
+                await api.destroy(`/crosslinks/${props.crosslink.id}/vote`, user?.token);
+            } catch (err) {
+                console.log(err);
+            } finally {
+                return;
+            }
+        } else if (changingVote) {
+            if (vote) {
+                setUserHasUpvoted(true);
+                setUpvotes((prev) => prev + 1);
+                if (downToUp) {
+                    setUserHasDownVoted(false);
+                    setDownvotes((prev) => prev - 1);
+                }
+            } else {
+                setUserHasDownVoted(true);
+                setDownvotes((prev) => prev + 1);
+                if (upToDown) {
+                    setUserHasUpvoted(false);
+                    setUpvotes((prev) => prev - 1);
+                }
+            }
+            try {
+                await api.put(`/crosslinks/${props.crosslink.id}/vote`, { vote }, user?.token);
```

Comment:
For voting I wrote the API end as an upsert, so you can use the same endpoint whenever you're casting a vote (creates/updates both use the PUT). I think it keeps it simpler on the API side not having to work out the reasoning behind the change, just making it.

Reply:
AFAIK there's no commitment to following REST exactly. Definitely open to doing it that way if it's better but I don't understand the advantage of having to delete the vote before creating a new one when it could be done in one call.